BorderBoxControl: Improve accessibility and labelling

aaronrobertshaw opened the pull request:
Related:
- https://github.com/WordPress/gutenberg/issues/42095
- https://github.com/WordPress/gutenberg/pull/42348

## What?

Updates the `BorderBoxControl` component to improve labelling, DOM Structure etc.

## Why?

Accessibility is important.

## How?
- Changes the control's wrapping element to a `fieldset` and its primary label to a `legend`
- Updates calls to `getByRole()` in the unit tests to also filter inputs by name.

## Testing Instructions
1. Run `npm run test-unit /packages/components/src/border-box-control/test`
2. Double check the `BorderBoxControl` continues to function correctly within the Storybook
    - http://localhost:50240/?path=/story/components-experimental-borderboxcontrol--default
3. Inspect the storybook control example and confirm the control is wrapped in a `fieldset` with an appropriate `legend` as the label.
4. Edit a post within the block or site editors, add a group block, select it, and then check the border controls still work

aaronrobertshaw committed:
Make BorderBoxControl wrapper a fieldset and label a legend

aaronrobertshaw committed:
Update getByRole to also filter by name

aristath commented:
Thank you for this PR @aaronrobertshaw! cc @afercia since you were checking these labels last week 👍

aaronrobertshaw committed:
Switch BorderBoxControl to use StyledLabel as legend element

ciampo reviewed (commented):
I wonder if, given the recent changes to `BorderControl`, is it going to be ok to have nested `fieldset`s. I have looked around for answers, and only found a few pointers.

> It is possible to nest one `fieldset` element inside another, but it is not recommended. Screen readers do not automatically indicate the end of the `fieldset` element, so it is impossible for screen reader users to confidently know which fields belong within which fieldset.

Source https://accessibility.blog.gov.uk/2016/07/22/using-the-fieldset-and-legend-elements/

Also found a few interesting related comments in https://www.tpgi.com/fieldsets-legends-and-screen-readers/


An alternative approach, according to the [official W3C's guidelines](https://www.w3.org/WAI/tutorials/forms/grouping/), would be to use a `div role=group` and the `aria-labelledby` attribute — @afercia , do you have any insights?

aaronrobertshaw committed:
Avoid nested fieldsets and use role=group approach

aaronrobertshaw committed:
Add changelog

aaronrobertshaw commented:
Thanks for the link to the [alternate approach](https://www.w3.org/WAI/tutorials/forms/grouping/) in the W3c Guidelines @ciampo 🙇 

I've updated this PR to adopt the `div[role="group"]` & `aria-labelledby` approach.

ciampo commented:
It would be great if @afercia could help us validate this new approach.

If the approach is validated, we could even consider using this approach for BorderControl, to avoid any potential nested fieldset/legend in higher level components?

afercia commented:
Thanks for working on this and for the ping.
I'd agree nesting fieldset elements is not recommended. It may make sense for large, complex, forms but that's not our case. Also, nesting fieldset elements impacts the way screen readers announce the group.

The post on accessibility.blog.gov.uk is a very good and informative source. It's the page I usually point people to for fieldset and legend related questions. However, the statement:

> Screen readers do not automatically indicate the end of the fieldset element

isn't entirely accurate. Actually, some screen readers do announce the end of a fieldset, some don't. I'm assuming that statement mostly refers to the case of _nested_ fieldsets.

> An alternative approach, according to the official W3C's guidelines, would be to use a div role=group and the aria-labelledby attribute

Not surprisingly, the way screen readers announce this pattern may be slightly different from the way they announce a HTML fieldset + legend. In my testing, some screen readers repeat the legend when focusing each input field within the fieldset. They don't do this with the ARIA pattern.

While this would be a minor inconsistency, the golden rule we should follow is to _always_ use HTML features when possible and avoid unnecessary ARIA. See the [First Rule of ARIA Use](https://www.w3.org/TR/using-aria/#rule1).

I haven't looked at the code so I'm not sure why there would be nested fieldsets but if possible I'd recommend to make a little effort to stick with HTML features and use only one fieldset + legend.

ciampo commented:
> I haven't looked at the code so I'm not sure why there would be nested fieldsets but if possible I'd recommend to make a little effort to stick with HTML features and use only one fieldset + legend.

In the context of https://github.com/WordPress/gutenberg/issues/42095, we've been working on making improvements to both `BorderControl` and `BorderBoxControl` by using the `legend` and `fieldset` HTML elements, following the advice received in https://github.com/WordPress/gutenberg/pull/42118#issuecomment-1175921129.

Such improvements have already been applied to `BorderControl` in https://github.com/WordPress/gutenberg/pull/42348, while this PR aims at applying them to `BorderBoxControl`.

The thing is, `BorderBoxControl` internally uses `BorderControl` — hence why we would end up in a situation where we have nested `fieldset` elements.

Given the extra context, do you have any advice on what the best solution would be?

tellthemachines commented:
Thanks for working on this!

> Not surprisingly, the way screen readers announce this pattern may be slightly different from the way they announce a HTML fieldset + legend. In my testing, some screen readers repeat the legend when focusing each input field within the fieldset. They don't do this with the ARIA pattern.

Yeah, a quick test of the storybook components on this branch with VoiceOver shows the label for the `role="group"` element isn't being read out when entering the component by tabbing or when accessing it through the form controls menu; funnily enough if we Shift+Tab in, it does get read 🤔 (For comparison, VoiceOver reads the `legend` of the neighbouring BorderControl component when entering each of its inputs.)

Could we perhaps make BorderControl render a fieldset by default, but allow it to render as something else with a prop? That would allow us to wrap BorderBoxControl in a fieldset with a legend, while wrapping the BorderControl inside it in a plain div. Maybe just a boolean prop that removes the semantic wrapper and the legend, allowing the form fields to be used as part of a larger semantic grouping?

Unrelated to this PR, but the "Border color and style picker" button is extremely verbose when non-default styles are selected. Are there already planned improvements to that or should I open an issue?

afercia commented:
@ciampo thanks for the clarification. Looking a bit into the code, I'm not sure adding the fieldset + legend to `BorderControl` is the best option. For example, when clicking on 'Unlink sides' this will render 4 fieldsets/legends, one for each side:

<img width="334" alt="Screenshot 2022-07-19 at 10 10 20" src="https://user-images.githubusercontent.com/1682452/179703524-e76ec23e-9bdf-4263-a167-29c82a485d0c.png">

That's a bit too much 🙂 Ideally, there should be only one fieldset/legend wrapping all the 4 controls.

I'd tend to think adding the fieldset/legend to the `BorderBoxControl` would be a better option. In terms of HTML, I'd recommend to make this UI consistent with the Radius control, with only one fieldset/legend wrapping all the fieds.

Is `BorderControl` always used internally by some other component? In this case, fieldset/legend could be a responsibility of the wrapping component. Otherwise, maybe make it optional as @tellthemachines suggested.

Aside: does anyone know what the `label` prop of `ToolsPanelItem` is supposed to do? I'm not sure it does anything.

ciampo commented:
> Is `BorderControl` always used internally by some other component? In this case, fieldset/legend could be a responsibility of the wrapping component. Otherwise, maybe make it optional as @tellthemachines suggested.

I'll let other folks who are more knowledgeable about how `BorderControl` is used across the block editor answer to this question — but in general, I agree with the statement above.

> Unrelated to this PR, but the "Border color and style picker" button is extremely verbose when non-default styles are selected. Are there already planned improvements to that or should I open an issue?

No planned improvements, AFAIK (cc'ing @aaronrobertshaw )

> Aside: does anyone know what the label prop of ToolsPanelItem is supposed to do? I'm not sure it does anything.

From the component's README:

https://github.com/WordPress/gutenberg/blob/1620d517274476321fdbc65fc1eeca0ef985ad39/packages/components/src/tools-panel/tools-panel-item/README.md?plain=1#L36-L46

aaronrobertshaw commented:
> Is BorderControl always used internally by some other component? In this case, fieldset/legend could be a responsibility of the wrapping component. Otherwise, maybe make it optional as @tellthemachines suggested.

At present, the `BorderControl` is only used via the `BorderBoxControl` in the block and site editors. 

Unfortunately, I ran out of time today to update both this PR and the `BorderControl` but it is first on the list for tomorrow.

> Unrelated to this PR, but the "Border color and style picker" button is extremely verbose when non-default styles are selected. Are there already planned improvements to that or should I open an issue?

I'd be happy to change it to anything that is easier on the ear or eyes. I'm open to any suggestions that might strike the right balance between clarity of purpose and brevity. ⚖️

aaronrobertshaw committed:
Revert "Avoid nested fieldsets and use role=group approach"

This reverts commit cdf6a48f35ff2372907fe90d313d5c321bb0065a.

aaronrobertshaw commented:
Thanks everyone for all the feedback on this. 🙇 

Given the `BorderControl` is exported from the components package and could be used outside of a `BorderBoxControl`. I've gone ahead and updated the `BorderControl` to optionally render with a fieldset/legend combo (See https://github.com/WordPress/gutenberg/pull/42611).

As for the `BorderBoxControl` and this PR, I've reverted https://github.com/WordPress/gutenberg/pull/42351/commits/7b48963744e01e0fb47cabae116bcfc2b0837e90 so a `BorderBoxControl` will now render as a `fieldset` again with a `legend` for its label.

ciampo commented:
Sounds good! Let's focus on https://github.com/WordPress/gutenberg/pull/42611 before getting back to this PR

aaronrobertshaw commented:
Closing this PR along with https://github.com/WordPress/gutenberg/pull/42611 due to their age and lack of a clear solution for now.

The root issues (https://github.com/WordPress/gutenberg/issues/42095, https://github.com/WordPress/gutenberg/issues/42630) will remain open as they are still important to address when possible.
